sw/airborne/math/wls/wls_alloc.c

```diff
@@ -107,44 +102,44 @@ void qr_solve_wrapper(int m, int n, float** A, float* b, float* x) {
  * control vector (sqare root of gamma)
  * @param imax Max number of iterations
  *
- * @return Number of iterations
+ * @return Number of iterations, -1 upon failure
  */
-int wls_alloc(float* u, float* v, float* umin, float* umax, float** B,
-    float* u_guess, float* W_init, float* Wv, float* Wu, float* up,
-    float gamma_sq, int imax) {
+int wls_alloc(float* u, int n_u, float* v, int n_v,
+    float* umin, float* umax, float** B,
+    float* u_guess, float* W_init, float* Wv, float* Wu,
+    float* up, float gamma_sq, int imax)
+{
   // allocate variables, use defaults where parameters are set to 0
   if(!gamma_sq) gamma_sq = 100000;
   if(!imax) imax = 100;
 
-  int n_c = CA_N_C;
-  int n_u = CA_N_U;
-  int n_v = CA_N_V;
+  int n_c = n_u + n_v;
 
-  float A[CA_N_C][CA_N_U];
-  float A_free[CA_N_C][CA_N_U];
+  float A[n_c][n_u];
```

Comment:
Does this still work with older compilers?

Reply:
Not sure, but the question is if you want this in the first place. This creates a Variable Length Array, which creates a ton of overhead in the code to manage the pointers to the memory. Though possible, It has been called 'actively stupid' by some 😄 
@gautierhattenberger This was changed in #2523

Reply:
At least it compiles with gcc-arm-none-eabi 15:9-2019-q4-0ubuntu1 from Ubuntu 20.04, which is the oldest version officially supported. In the original code, only the size is static, right ? Maybe it helps for compiler optimization, but you still need to allocate on the stack at each call, so I'm not sure it makes a big difference.

Reply:
@EwoudSmeur hum, I forgot about these changes. Then, it can be restored and be part of the wls module configuration. The user will have to make sure the matrix is big enough for all cases for a given configuration.
Did you make performance tests back then ?

Reply:
I did not do explicit performance checks on this code, but after someone explained to me why VLAs create extra code I had a look at the machine code that is generated from a simple program with a VLA. It explodes quite quickly even for simple operations, The problem is not that the memory needs to be allocated, but that at compile time it is not clear where the memory is going to be. All kinds of clever pointer magic has to be done to keep track, not only of the VLA memory itself, but all the memory that comes after.